<a href="https://colab.research.google.com/github/amaan-zafar/data-science-python/blob/2-Proxy-Utilisation/Proxy_utilization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Importing required libraries
from lxml.html import fromstring
import requests
from itertools import cycle
import traceback

In [ ]:
def get_proxies():
  PROXY_URL = 'https://free-proxy-list.net/'
  response = requests.get(PROXY_URL)
  parser = fromstring(response.text)
  proxies = set()     
  for i in parser.xpath('//tbody/tr'):
    if i.xpath('.//td[7][contains(text(),"yes")]'):
      proxy = ":".join([i.xpath('.//td[1]/text()')[0], i.xpath('.//td[2]/text()')[0]])
      proxies.add(proxy)
  return proxies

proxies = get_proxies()
print(proxies)
proxy_len = len(proxies)
print('Number of proxies are :', proxy_len)
proxy_pool = cycle(proxies)
url = 'https://httpbin.org/ip'
proxy_success = 0
proxy_blocked = 0
for i in range(1,proxy_len + 1):
  #Get a proxy from the pool
  proxy = next(proxy_pool)
  print("Request #%d"%i)
  try:
    response = requests.get(url,proxies={"http": proxy, "https": proxy})
    print(response.status_code)
    proxy_success += 1
    if (response.status_code==403):
      proxy_blocked += 1
  except Exception as e:
    print("Error :", e.__class__, "occurred. Skipping.")
success_percentage = proxy_success / proxy_len * 100.0
print(proxy_success)
print(proxy_blocked)
print(success_percentage)

{'103.109.57.1:8080', '150.129.148.88:35101', '178.217.216.184:49086', '189.52.154.213:3128', '95.154.104.147:44393', '103.148.79.107:8080', '85.15.152.39:3128', '181.129.52.154:42648', '37.120.192.154:8080', '62.33.210.34:58918', '103.199.157.153:40049', '37.57.15.43:33761', '92.60.190.249:50335', '213.230.90.106:3128', '14.192.31.4:50838', '95.165.163.188:60103', '178.47.141.128:46321', '91.193.253.188:23500', '140.227.80.43:3128', '82.99.217.18:8080', '129.205.161.242:31075', '181.129.43.3:8080', '79.101.98.2:53281', '150.107.207.137:61954', '182.253.82.154:37242', '41.234.66.33:3128', '101.255.72.171:80', '188.120.235.217:8118', '213.230.97.10:3128', '101.32.191.99:22225', '101.96.116.89:8080', '45.236.152.45:6666', '115.86.65.121:8080', '93.170.118.249:8080', '113.160.224.224:55443', '200.29.109.112:44749', '202.180.54.97:8080', '103.145.45.6:55443', '37.18.100.185:8080', '103.9.188.229:36984', '94.130.219.212:3128', '201.132.105.30:8080', '103.215.177.153:8888', '13.235.137.56:80

**Stats and observations**
URL used : https://httpbin.org/ip

Total Proxy Used : 103

Total Successful Proxy attempts : 49

  Successful proxy attempts with Status 200 : 17
  
  Successful proxy attempts with Status 403 : 32

Total Proxy with Connection Error : 54

Percentage Successful : 47.57%

